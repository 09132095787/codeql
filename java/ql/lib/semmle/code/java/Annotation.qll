/**
 * Provides classes and predicates for working with Java annotations.
 *
 * Annotations are used to add meta-information to language elements in a
 * uniform fashion. They can be seen as typed modifiers that can take
 * parameters.
 *
 * Each annotation type has zero or more annotation elements that contain a
 * name and possibly a value.
 */

import Element
import Expr
import Type
import Member
import JDKAnnotations

/** Any annotation used to annotate language elements with meta-information. */
class Annotation extends @annotation, Expr {
  /** Holds if this annotation applies to a declaration. */
  predicate isDeclAnnotation() { this instanceof DeclAnnotation }

  /** Holds if this annotation applies to a type. */
  predicate isTypeAnnotation() { this instanceof TypeAnnotation }

  /** Gets the element being annotated. */
  Element getAnnotatedElement() {
    exists(Element e | e = this.getParent() |
      if e.(Field).getCompilationUnit().fromSource()
      then
        exists(FieldDeclaration decl |
          decl.getField(0) = e and
          result = decl.getAField()
        )
      else result = e
    )
  }

  /** Gets the annotation type declaration for this annotation. */
  override AnnotationType getType() { result = Expr.super.getType() }

  /** Gets the annotation element with the specified `name`. */
  AnnotationElement getAnnotationElement(string name) {
    result = this.getType().getAnnotationElement(name)
  }

  /**
   * DEPRECATED: Getting the value of _any_ annotation element is error-prone because
   * it could lead to selecting the value of the wrong element by accident (for example
   * when an annotation type is extended in the future). Prefer the predicate `getValue(string)`
   * and explicitly specify the element name. Use `getValue(_)` if it is really desired to
   * get the value of any element.
   *
   * Gets a value of an annotation element. This includes default values in case
   * no explicit value is specified. For elements with an array value type this
   * might have an `ArrayInit` as result. To properly handle array values, prefer
   * the predicate `getAnArrayValue`.
   */
  deprecated Expr getAValue() { filteredAnnotValue(this, _, result) }

  /**
   * Gets the value of the annotation element with the specified `name`.
   * This includes default values in case no explicit value is specified.
   * For elements with an array value type this might have an `ArrayInit` as result.
   * To properly handle array values, prefer the predicate `getAnArrayValue`.
   */
  Expr getValue(string name) { filteredAnnotValue(this, this.getAnnotationElement(name), result) }

  /**
   * If the value type of the annotation element with the specified `name` is an enum type,
   * gets the enum constant used as value for that element. This includes default values in
   * case no explicit value is specified.
   *
   * If the element value type is an enum type array, use `getAnEnumConstantArrayValue`.
   */
  EnumConstant getEnumConstantValue(string name) { result = getValue(name).(FieldRead).getField() }

  /**
   * If the value type of the annotation element with the specified `name` is `String`,
   * gets the string value used for that element. This includes default values in case no
   * explicit value is specified.
   *
   * If the element value type is a string array, use `getAStringArrayValue`.
   */
  string getStringValue(string name) {
    // Uses CompileTimeConstantExpr instead of StringLiteral because value can
    // be read of final variable as well
    result = getValue(name).(CompileTimeConstantExpr).getStringValue()
  }

  /**
   * If the value type of the annotation element with the specified `name` is `int`,
   * gets the int value used for that element. This includes default values in case no
   * explicit value is specified.
   *
   * If the element value type is an `int` array, use `getAnIntArrayValue`.
   */
  int getIntValue(string name) {
    // Uses CompileTimeConstantExpr instead of IntegerLiteral because value can
    // be read of final variable as well
    result = getValue(name).(CompileTimeConstantExpr).getIntValue()
  }

  /**
   * If the value type of the annotation element with the specified `name` is `boolean`,
   * gets the boolean value used for that element. This includes default values in case
   * no explicit value is specified.
   */
  boolean getBooleanValue(string name) {
    // Uses CompileTimeConstantExpr instead of BooleanLiteral because value can
    // be read of final variable as well
    result = getValue(name).(CompileTimeConstantExpr).getBooleanValue()
  }

  /**
   * If the annotation element with the specified `name` has a Java `Class` as value type,
   * gets the referenced type used as value for that element. This includes default values
   * in case no explicit value is specified.
   *
   * If the element value type is a `Class` array, use `getATypeArrayValue`.
   */
  Type getTypeValue(string name) { result = getValue(name).(TypeLiteral).getReferencedType() }

  /** Gets the element being annotated. */
  Element getTarget() { result = this.getAnnotatedElement() }

  override string toString() { result = this.getType().getName() }

  /** This expression's Halstead ID (used to compute Halstead metrics). */
  override string getHalsteadID() { result = "Annotation" }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an array
   * type. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression defined for the value.
   */
  Expr getAnArrayValue(string name) { result = getArrayValue(name, _) }

  /**
   * DEPRECATED: Predicate has been renamed to `getAnArrayValue`
   */
  deprecated Expr getAValue(string name) { result = getAnArrayValue(name) }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an enum
   * type array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression defined for the value.
   */
  EnumConstant getAnEnumConstantArrayValue(string name) {
    result = this.getAnArrayValue(name).(FieldRead).getField()
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as a string
   * array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression defined for the value.
   */
  string getAStringArrayValue(string name) {
    result = this.getAnArrayValue(name).(CompileTimeConstantExpr).getStringValue()
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an `int`
   * array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression defined for the value.
   */
  int getAnIntArrayValue(string name) {
    result = this.getAnArrayValue(name).(CompileTimeConstantExpr).getIntValue()
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as a `Class`
   * array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression defined for the value.
   */
  Type getATypeArrayValue(string name) {
    result = this.getAnArrayValue(name).(TypeLiteral).getReferencedType()
  }

  /**
   * Gets the value at a given index of the annotation element with the specified `name`, which must be
   * declared as an array type. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be the element
   * at the given index of that array. Otherwise, the result will be the single expression defined for
   * the value and the `index` will be 0.
   */
  Expr getArrayValue(string name, int index) {
    this.getType().getAnnotationElement(name).getType() instanceof Array and
    exists(Expr value | value = this.getValue(name) |
      if value instanceof ArrayInit
      then result = value.(ArrayInit).getInit(index)
      else (
        index = 0 and result = value
      )
    )
  }

  override string getAPrimaryQlClass() { result = "Annotation" }
}

/** An `Annotation` that applies to a declaration. */
class DeclAnnotation extends @declannotation, Annotation { }

/** An `Annotation` that applies to a type. */
class TypeAnnotation extends @typeannotation, Annotation { }

/**
 * There may be duplicate entries in annotValue(...) - one entry for
 * information populated from bytecode, and one for information populated
 * from source. This removes the duplication.
 */
private predicate filteredAnnotValue(Annotation a, Method m, Expr val) {
  annotValue(a, m, val) and
  (sourceAnnotValue(a, m, val) or not sourceAnnotValue(a, m, _))
}

private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
  annotValue(a, m, val) and
  val.getFile().getExtension() = "java"
}

/** An abstract representation of language elements that can be annotated. */
class Annotatable extends Element {
  /** Holds if this element has an annotation, including inherited annotations. */
  predicate hasAnnotation() { exists(getAnAnnotation()) }

  /** Holds if this element has a declared annotation, excluding inherited annotations. */
  predicate hasDeclaredAnnotation() { exists(getADeclaredAnnotation()) }

  /**
   * Holds if this element has the specified annotation, including inherited
   * annotations.
   */
  predicate hasAnnotation(string package, string name) {
    exists(AnnotationType at | at = this.getAnAnnotation().getType() |
      at.nestedName() = name and at.getPackage().getName() = package
    )
  }

  /**
   * Gets an annotation that applies to this element, including inherited annotations.
   * The results only include _direct_ annotations; _indirect_ annotations, that is
   * repeated annotations in an (implicit) container annotation, are not included.
   */
  // This predicate is overridden by Class to consider inherited annotations
  cached
  Annotation getAnAnnotation() { result = getADeclaredAnnotation() }

  /**
   * Gets an annotation that is declared on this element, excluding inherited annotations.
   */
  Annotation getADeclaredAnnotation() { result.getAnnotatedElement() = this }

  /** Gets an _indirect_ (= repeated) annotation. */
  // 'indirect' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
  private Annotation getAnIndirectAnnotation() {
    exists(AnnotationType t, Annotation containerAnn |
      t = result.getType() and
      containerAnn = getADeclaredAnnotation() and
      containerAnn.getType() = t.getContainingAnnotationType()
    |
      result = containerAnn.getAnArrayValue("value")
    )
  }

  private Annotation getADeclaredAssociatedAnnotation(AnnotationType t) {
    // Direct or indirect annotation
    result.getType() = t and result = [getADeclaredAnnotation(), getAnIndirectAnnotation()]
  }

  private Annotation getAnAssociatedAnnotation(AnnotationType t) {
    if exists(getADeclaredAssociatedAnnotation(t))
    then result = getADeclaredAssociatedAnnotation(t)
    else (
      // Only if neither a direct nor an indirect annotation is present look for an inherited one
      t.isInherited() and
      // @Inherited only works for classes; cast to Annotatable is necessary because predicate is private
      result = this.(Class).getASupertype().(Class).(Annotatable).getAnAssociatedAnnotation(t)
    )
  }

  /**
   * Gets an annotation _associated_ with this element, that is:
   * - An annotation directly present on this element, or
   * - An annotation indirectly present on this element (in the form of a repeated annotation), or
   * - If an annotation of a type is neither directly nor indirectly present
   *   the result is an associated inherited annotation (recursively)
   */
  // 'associated' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
  Annotation getAnAssociatedAnnotation() { result = getAnAssociatedAnnotation(_) }

  /**
   * Holds if this or any enclosing `Annotatable` has a `@SuppressWarnings("<category>")`
   * annotation attached to it for the specified `category`.
   */
  predicate suppressesWarningsAbout(string category) {
    category = this.getAnAnnotation().(SuppressWarningsAnnotation).getASuppressedWarning()
    or
    this.(Member).getDeclaringType().suppressesWarningsAbout(category)
    or
    this.(Expr).getEnclosingCallable().suppressesWarningsAbout(category)
    or
    this.(Stmt).getEnclosingCallable().suppressesWarningsAbout(category)
    or
    this.(NestedClass).getEnclosingType().suppressesWarningsAbout(category)
    or
    this.(LocalVariableDecl).getCallable().suppressesWarningsAbout(category)
  }
}

/** An annotation type is a special kind of interface type declaration. */
class AnnotationType extends Interface {
  AnnotationType() { isAnnotType(this) }

  /** Gets the annotation element with the specified `name`. */
  AnnotationElement getAnnotationElement(string name) {
    methods(result, _, _, _, this, _) and result.hasName(name)
  }

  /** Gets an annotation element that is a member of this annotation type. */
  AnnotationElement getAnAnnotationElement() { methods(result, _, _, _, this, _) }

  /** Holds if this annotation type is annotated with the meta-annotation `@Inherited`. */
  predicate isInherited() {
    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Inherited")
  }

  /** Holds if this annotation type is annotated with the meta-annotation `@Documented`. */
  predicate isDocumented() {
    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Documented")
  }

  /**
   * Gets the retention policy of this annotation type, that is, the name of one of the
   * enum constants of `java.lang.annotation.RetentionPolicy`. If no explicit retention
   * policy is specified the result is `CLASS`.
   */
  string getRetentionPolicy() {
    if getADeclaredAnnotation() instanceof RetentionAnnotation
    then result = getADeclaredAnnotation().(RetentionAnnotation).getRetentionPolicy()
    else
      // If not explicitly specified retention is CLASS
      result = "CLASS"
  }

  /**
   * Holds if the element type is a possible target for this annotation type.
   * The `elementType` is the name of one of the `java.lang.annotation.ElementType`
   * enum constants. If no explicit target is specified for this annotation type
   * it is considered to be applicable to all elements.
   */
  // Note: Cannot use a predicate with string as result because annotation type without
  // explicit @Target can be applied to all targets, requiring to hardcode element types here
  bindingset[elementType]
  predicate isATargetType(string elementType) {
    if getADeclaredAnnotation() instanceof TargetAnnotation
    then elementType = getADeclaredAnnotation().(TargetAnnotation).getATargetElementType()
    else
      // No Target annotation means "applicable to all contexts" since JDK 14, see https://bugs.openjdk.java.net/browse/JDK-8231435
      // The compiler does not completely implement that, but pretend it did
      any()
  }

  /** Holds if this annotation type is annotated with the meta-annotation `@Repeatable`. */
  predicate isRepeatable() { getADeclaredAnnotation() instanceof RepeatableAnnotation }

  /**
   * If this annotation type is annotated with the meta-annotation `@Repeatable`,
   * gets the annotation type which acts as _containing annotation type_.
   */
  AnnotationType getContainingAnnotationType() {
    result = getADeclaredAnnotation().(RepeatableAnnotation).getContainingType()
  }
}

/** An annotation element is a member declared in an annotation type. */
class AnnotationElement extends Member {
  AnnotationElement() { isAnnotElem(this) }

  /** Gets the type of this annotation element. */
  Type getType() { methods(this, _, _, result, _, _) }

  /** Gets the Kotlin type of this annotation element. */
  KotlinType getKotlinType() { methodsKotlinType(this, result) }
}
