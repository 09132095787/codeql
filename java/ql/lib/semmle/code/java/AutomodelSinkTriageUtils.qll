private import java
private import semmle.code.java.dataflow.ExternalFlow as ExternalFlow
private import semmle.code.java.dataflow.TaintTracking::TaintTracking

/** Gets the models-as-data description for the method argument with the index `index`. */
bindingset[index]
private string getArgumentForIndex(int index) {
  index = -1 and result = "Argument[this]"
  or
  index >= 0 and result = "Argument[" + index + "]"
}

private boolean considerSubtypes(Callable callable) {
  if
    callable.isStatic() or
    callable.getDeclaringType().isStatic() or
    callable.isFinal() or
    callable.getDeclaringType().isFinal()
  then result = false
  else result = true
}

private class SinkModelExpr extends Expr {
  predicate hasSignature(
    string package, string type, boolean subtypes, string name, string signature, string input
  ) {
    exists(Call call, Callable callable, int argIdx |
      call.getCallee() = callable and
      (
        this = call.getArgument(argIdx)
        or
        this = call.getQualifier() and argIdx = -1
      ) and
      input = getArgumentForIndex(argIdx) and
      package = callable.getDeclaringType().getPackage().getName() and
      type = callable.getDeclaringType().getErasure().(RefType).nestedName() and
      subtypes = considerSubtypes(callable) and
      name = callable.getName() and
      signature = ExternalFlow::paramsString(callable)
    )
  }
}

private string pyBool(boolean b) {
  b = true and result = "True"
  or
  b = false and result = "False"
}

/**
 * Gets a string representation of the existing sink model at the expression `e`, in the format in
 * which it would appear in a Models-as-Data file.
 */
string getSinkModelRepr(SinkModelExpr e) {
  exists(
    string package, string type, boolean subtypes, string name, string signature, string input,
    string ext, string kind, string provenance
  |
    e.hasSignature(package, type, subtypes, name, signature, input) and
    ExternalFlow::sinkModel(package, type, subtypes, name, signature, ext, input, kind, provenance) and
    provenance = "ai-generated" and
    result =
      "\"" + package + "\", \"" + type + "\", " + pyBool(subtypes) + ", \"" + name + "\", \"" +
        signature + "\", \"" + ext + "\", \"" + input + "\", \"" + kind + "\", \"" + provenance +
        "\""
  )
}

/**
 * Gets the string representation of a sink model in a format suitable for appending to an alert
 * message.
 */
string getSinkModelQueryRepr(SinkModelExpr e) { result = "\nsinkModel: " + getSinkModelRepr(e) }
// TODO: make this logic more generic
// private predicate relevantSinkModel(int c, string s) {
//   exists(RequestForgeryFlow::PathNode source, RequestForgeryFlow::PathNode sink |
//     RequestForgeryFlow::flowPath(source, sink) and
//     s = getSinkModelRepr(sink.getNode().asExpr())
//   ) and
//   c =
//     count(RequestForgeryFlow::PathNode sink |
//       exists(RequestForgeryFlow::PathNode source |
//         RequestForgeryFlow::flowPath(source, sink) and
//         s = getSinkModelRepr(sink.getNode().asExpr())
//       )
//     )
// }
