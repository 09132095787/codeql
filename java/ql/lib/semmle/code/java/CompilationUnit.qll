/**
 * Provides classes and predicates for working with Java compilation units.
 */

import Element
import Package
import semmle.code.FileSystem

/**
 * A compilation unit is a `.java` or `.class` file.
 */
class CompilationUnit extends Element, File {
  CompilationUnit() { cupackage(this, _) }

  /** Gets the name of the compilation unit (not including its extension). */
  override string getName() { result = Element.super.getName() }

  /**
   * Holds if this compilation unit has the specified `name`,
   * which must not include the file extension.
   */
  override predicate hasName(string name) { Element.super.hasName(name) }

  override string toString() { result = Element.super.toString() }

  /** Gets the declared package of this compilation unit. */
  Package getPackage() { cupackage(this, result) }

  /**
   * Gets the module associated with this compilation unit, if any.
   */
  Module getModule() { cumodule(this, result) }

  /**
   * Gets a type which is available by its simple name in this compilation unit.
   * Reasons for this can be:
   * - The type is declared in this compilation unit as top-level type
   * - The type is imported
   * - The type is declared in the same package as this compilation unit
   * - The type is declared in the package `java.lang`
   */
  ClassOrInterface getATypeAvailableBySimpleName() {
    // See "Shadowing", https://docs.oracle.com/javase/specs/jls/se17/html/jls-6.html#jls-6.4.1
    // Note: Currently the logic below does not consider shadowing and might have multiple results
    // with the same type name
    result.(TopLevelType).getCompilationUnit() = this
    or
    exists(ImportStaticTypeMember importDecl |
      importDecl.getCompilationUnit() = this and
      result = importDecl.getATypeImport()
    )
    or
    exists(ImportType importDecl |
      importDecl.getCompilationUnit() = this and
      result = importDecl.getImportedType()
    )
    or
    exists(ImportStaticOnDemand importDecl |
      importDecl.getCompilationUnit() = this and
      result = importDecl.getATypeImport()
    )
    or
    exists(ImportOnDemandFromType importDecl |
      importDecl.getCompilationUnit() = this and
      result = importDecl.getAnImport()
    )
    or
    exists(ImportOnDemandFromPackage importDecl |
      importDecl.getCompilationUnit() = this and
      result = importDecl.getAnImport()
    )
    or
    // From same package or java.lang, see https://docs.oracle.com/javase/specs/jls/se17/html/jls-7.html
    result.(TopLevelType).getPackage() = this.getPackage()
    or
    result.(TopLevelType).getPackage().hasName("java.lang")
  }

  override string getAPrimaryQlClass() { result = "CompilationUnit" }
}
