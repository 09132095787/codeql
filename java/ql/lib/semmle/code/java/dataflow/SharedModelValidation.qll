/**
 * INTERNAL: Do not use.
 *
 * Provides classes for validating kinds in models as data rows.
 * Such that we can share this logic across our CodeQL analysis of different languages.
 */

/** A valid models-as-data sink kind. */
class ValidSinkKind extends string {
  ValidSinkKind() {
    this =
      [
        // shared ALL languages
        "request-forgery", "ldap-injection", "sql-injection", "nosql-injection", "log-injection",
        "xpath-injection", "html-injection", "js-injection", "url-redirection", "path-injection",
        "file-content-store", "hostname-verification", "response-splitting", "information-leak",
        "xslt-injection", "template-injection", "fragment-injection", "command-injection",
        "unsafe-deserialization", "xxe", "database-store", "format-string",
        // .matches("credentials-%"), .matches("regex-use%")"
        // shared MOST languages
        "code-injection", // .matches("encryption-%"),
        // Java only
        "jndi-injection", "mvel-injection", "groovy-injection", "ognl-injection", "jexl-injection",
        "bean-validation", "intent-redirection", "pending-intents",
        // JS only
        "mongodb.sink",
        // Swift only
        "preferences-store", "transmission", "predicate-injection", "webview-fetch",
        "tls-protocol-version", "hash-iteration-count" // .matches("%string-%length"), .matches("weak-hash-input-")
      ]
  }
}

/** A valid models-as-data source kind. */
class ValidSourceKind extends string {
  ValidSourceKind() {
    this =
      [
        // shared ALL languages
        "remote", "local"
      ]
  }
}

/** A valid models-as-data summary kind. */
class ValidSummaryKind extends string {
  ValidSummaryKind() {
    this =
      [
        // shared ALL languages
        "taint", "value"
      ]
  }
}

/** A valid models-as-data neutral kind. */
class ValidNeutralKind extends string {
  ValidNeutralKind() {
    this =
      [
        // shared ALL languages
        "summary", "source", "sink"
      ]
  }
}
