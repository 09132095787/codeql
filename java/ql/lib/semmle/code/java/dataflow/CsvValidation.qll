/** Provides a query predicate to check the CSV data for validation errors. */

import java
private import ExternalFlow
private import internal.FlowSummaryImpl::Private::External
private import internal.AccessPathSyntax

private string getInvalidModelInput() {
  exists(string pred, string input, string part |
    sinkModel(_, _, _, _, _, _, input, _, _) and pred = "sink"
    or
    summaryModel(_, _, _, _, _, _, input, _, _, _) and pred = "summary"
  |
    (
      invalidSpecComponent(input, part) and
      not part = "" and
      not (part = "Argument" and pred = "sink") and
      not parseArg(part, _)
      or
      part = input.(AccessPath).getToken(0) and
      parseParam(part, _)
    ) and
    result = "Unrecognized input specification \"" + part + "\" in " + pred + " model."
  )
}

private string getInvalidModelOutput() {
  exists(string pred, string output, string part |
    sourceModel(_, _, _, _, _, _, output, _, _) and pred = "source"
    or
    summaryModel(_, _, _, _, _, _, _, output, _, _) and pred = "summary"
  |
    invalidSpecComponent(output, part) and
    not part = "" and
    not (part = ["Argument", "Parameter"] and pred = "source") and
    result = "Unrecognized output specification \"" + part + "\" in " + pred + " model."
  )
}

private string getInvalidModelKind() {
  exists(string row, string kind | summaryModel(row) |
    kind = row.splitAt(";", 8) and
    not kind = ["taint", "value"] and
    result = "Invalid kind \"" + kind + "\" in summary model."
  )
  or
  exists(string row, string kind | sinkModel(row) |
    kind = row.splitAt(";", 7) and
    not kind =
      [
        "open-url", "jndi-injection", "ldap", "sql", "jdbc-url", "logging", "mvel", "xpath",
        "groovy", "xss", "ognl-injection", "intent-start", "pending-intent-sent", "url-open-stream",
        "url-redirect", "create-file", "write-file", "set-hostname-verifier", "header-splitting",
        "information-leak", "xslt", "jexl", "bean-validation"
      ] and
    not kind.matches("regex-use%") and
    not kind.matches("qltest%") and
    result = "Invalid kind \"" + kind + "\" in sink model."
  )
  or
  exists(string row, string kind | sourceModel(row) |
    kind = row.splitAt(";", 7) and
    not kind = ["remote", "contentprovider", "android-widget", "android-external-storage-dir"] and
    not kind.matches("qltest%") and
    result = "Invalid kind \"" + kind + "\" in source model."
  )
}

private string getInvalidModelSubtype() {
  exists(string pred, string row, int expect |
    sourceModel(row) and expect = 9 and pred = "source"
    or
    sinkModel(row) and expect = 9 and pred = "sink"
    or
    summaryModel(row) and expect = 10 and pred = "summary"
  |
    exists(string b |
      b = row.splitAt(";", 2) and
      not b = ["true", "false"] and
      result = "Invalid boolean \"" + b + "\" in " + pred + " model."
    )
  )
}

private string getInvalidModelColumnCount() {
  exists(string pred, string row, int expect |
    sourceModel(row) and expect = 9 and pred = "source"
    or
    sinkModel(row) and expect = 9 and pred = "sink"
    or
    summaryModel(row) and expect = 10 and pred = "summary"
    or
    negativeSummaryModel(row) and expect = 5 and pred = "negative summary"
  |
    exists(int cols |
      cols = 1 + max(int n | exists(row.splitAt(";", n))) and
      cols != expect and
      result =
        "Wrong number of columns in " + pred + " model row, expected " + expect + ", got " + cols +
          " in " + row + "."
    )
  )
}

private string getInvalidModelSignature() {
  exists(
    string pred, string namespace, string type, string name, string signature, string ext,
    string provenance
  |
    sourceModel(namespace, type, _, name, signature, ext, _, _, provenance) and pred = "source"
    or
    sinkModel(namespace, type, _, name, signature, ext, _, _, provenance) and pred = "sink"
    or
    summaryModel(namespace, type, _, name, signature, ext, _, _, _, provenance) and
    pred = "summary"
    or
    negativeSummaryModel(namespace, type, name, signature, provenance) and
    ext = "" and
    pred = "nonesummary"
  |
    not namespace.regexpMatch("[a-zA-Z0-9_\\.]+") and
    result = "Dubious namespace \"" + namespace + "\" in " + pred + " model."
    or
    not type.regexpMatch("[a-zA-Z0-9_\\$<>]+") and
    result = "Dubious type \"" + type + "\" in " + pred + " model."
    or
    not name.regexpMatch("[a-zA-Z0-9_]*") and
    result = "Dubious name \"" + name + "\" in " + pred + " model."
    or
    not signature.regexpMatch("|\\([a-zA-Z0-9_\\.\\$<>,\\[\\]]*\\)") and
    result = "Dubious signature \"" + signature + "\" in " + pred + " model."
    or
    not ext.regexpMatch("|Annotated") and
    result = "Unrecognized extra API graph element \"" + ext + "\" in " + pred + " model."
    or
    not provenance = ["manual", "generated"] and
    result = "Unrecognized provenance description \"" + provenance + "\" in " + pred + " model."
  )
}

/** Holds if some row in a CSV-based flow model appears to contain typos. */
query predicate invalidModelRow(string msg) {
  msg =
    [
      getInvalidModelSignature(), getInvalidModelInput(), getInvalidModelOutput(),
      getInvalidModelSubtype(), getInvalidModelColumnCount(), getInvalidModelKind()
    ]
}
