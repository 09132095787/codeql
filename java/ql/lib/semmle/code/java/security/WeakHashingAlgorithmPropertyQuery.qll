/** Provides classes and predicates to reason about property files and weak hashing algorithms. */

import java
private import semmle.code.configfiles.ConfigFiles
private import semmle.code.java.dataflow.DataFlow
private import semmle.code.java.dataflow.TaintTracking
private import semmle.code.java.security.Encryption
private import semmle.code.java.frameworks.Properties
private import semmle.code.java.dataflow.RangeUtils

private class GetPropertyMethodCall extends MethodCall {
  GetPropertyMethodCall() { this.getMethod() instanceof PropertiesGetPropertyMethod }

  private ConfigPair getPair() {
    this.getArgument(0).(ConstantStringExpr).getStringValue() = result.getNameElement().getName()
  }

  string getPropertyValue() {
    result = this.getPair().getValueElement().getValue() or
    result = this.getArgument(1).(ConstantStringExpr).getStringValue()
  }
}

/**
 * Get the name of the weak cryptographic algorithm represented by `node`.
 */
string getWeakHashingAlgorithmName(DataFlow::Node node) {
  exists(MethodCall mc, ConfigPair pair |
    node.asExpr() = mc and mc.getMethod() instanceof PropertiesGetPropertyMethod
  |
    mc.getArgument(0).(ConstantStringExpr).getStringValue() = pair.getNameElement().getName() and
    pair.getValueElement().getValue() = result and
    not pair.getValueElement().getValue().regexpMatch(getSecureAlgorithmRegex())
  )
}

/**
 * Dataflow configuration from a configuration pair in a properties file to the use of a cryptographic algorithm.
 */
module InsecureAlgorithmPropertyConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node n) {
    exists(GetPropertyMethodCall mc, string algo | n.asExpr() = mc |
      algo = mc.getPropertyValue() and
      not algo.regexpMatch(getSecureAlgorithmRegex())
    )
  }

  predicate isSink(DataFlow::Node n) { n.asExpr() = any(CryptoAlgoSpec c).getAlgoSpec() }
}

/**
 * Dataflow from a configuration pair in a properties file to the use of a cryptographic algorithm.
 */
module InsecureAlgorithmPropertyFlow = TaintTracking::Global<InsecureAlgorithmPropertyConfig>;
