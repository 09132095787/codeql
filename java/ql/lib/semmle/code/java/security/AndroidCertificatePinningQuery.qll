/** Definitiona for the Android Missing Certificate Pinning query. */

import java
import semmle.code.xml.AndroidManifest
import semmle.code.java.dataflow.TaintTracking
import HttpsUrls

class AndroidNetworkSecurityConfigFile extends XmlFile {
  AndroidNetworkSecurityConfigFile() {
    exists(AndroidApplicationXmlElement app, AndroidXmlAttribute confAttr, string confName |
      confAttr.getElement() = app and
      confAttr.getValue() = "@xml/" + confName and
      this.getRelativePath().matches("%res/xml/" + confName + ".xml") and
      this.getARootElement().getName() = "network-security-config"
    )
  }
}

predicate isAndroid() { exists(AndroidManifestXmlFile m) }

predicate trustedDomain(string domainName) {
  exists(
    AndroidNetworkSecurityConfigFile confFile, XmlElement domConf, XmlElement domain,
    XmlElement trust
  |
    domConf.getFile() = confFile and
    domConf.getName() = "domain-config" and
    domain.getParent() = domConf and
    domain.getName() = "domain" and
    domain.getACharactersSet().getCharacters() = domainName and
    trust.getParent() = domConf and
    trust.getName() = ["trust-anchors", "pin-set"]
  )
}

private class UntrustedUrlConfig extends TaintTracking::Configuration {
  UntrustedUrlConfig() { this = "UntrustedUrlConfig" }

  override predicate isSource(DataFlow::Node node) {
    exists(string d | trustedDomain(d)) and
    exists(string lit | lit = node.asExpr().(CompileTimeConstantExpr).getStringValue() |
      lit.matches("%://%") and // it's a URL
      not exists(string dom | trustedDomain(dom) and lit.matches("%" + dom + "%"))
    )
  }

  override predicate isSink(DataFlow::Node node) { node instanceof UrlOpenSink }
}

predicate missingPinning(DataFlow::Node node) {
  isAndroid() and
  node instanceof UrlOpenSink and
  (
    not exists(string s | trustedDomain(s))
    or
    exists(UntrustedUrlConfig conf | conf.hasFlow(_, node))
  )
}
