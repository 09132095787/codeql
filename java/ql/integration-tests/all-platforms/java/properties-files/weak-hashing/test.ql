import java
import semmle.code.java.dataflow.DataFlow
import semmle.code.java.security.WeakHashingAlgorithmPropertyQuery

query predicate weakAlgorithmUse(DataFlow::Node sink) {
  exists(DataFlow::Node source | InsecureAlgorithmPropertyFlow::flow(source, sink))
}
