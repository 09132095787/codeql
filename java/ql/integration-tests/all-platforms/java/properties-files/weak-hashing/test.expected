| src/main/java/com/example/WeakHashing.java:14:55:14:83 | getProperty(...) |